File: src/brs/crypto/Crypto.java
package brs.crypto;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import org.bouncycastle.jcajce.provider.digest.MD5;

import signumj.crypto.SignumCrypto;
import signumj.entity.SignumID;

public final class Crypto {
  static final SignumCrypto signumCrypto = SignumCrypto.getInstance();

  private Crypto() {
  } //never

  public static MessageDigest sha256() {
    return signumCrypto.getSha256();
  }

  public static MessageDigest shabal256() {
    return signumCrypto.getShabal256();
  }

  public static MessageDigest ripemd160() {
    return signumCrypto.getRipeMD160();
  }

  public static MessageDigest md5() {// TODO unit test
    try {
      return MessageDigest.getInstance("MD5"); // TODO burstkit4j integration
    } catch (NoSuchAlgorithmException e) {
      return new MD5.Digest();
    }
  }

  public static byte[] getPublicKey(String secretPhrase) {
    return signumCrypto.getPublicKey(secretPhrase);
  }

  public static byte[] getPrivateKey(String secretPhrase) {
    return signumCrypto.getPrivateKey(secretPhrase);
  }

  public static byte[] sign(byte[] message, String secretPhrase) {
      return signumCrypto.sign(message, secretPhrase);
  }

  public static boolean verify(byte[] signature, byte[] message, byte[] publicKey, boolean enforceCanonical) {
      return signumCrypto.verify(signature, message, publicKey, enforceCanonical);
  }

  public static byte[] aesEncrypt(byte[] plaintext, byte[] myPrivateKey, byte[] theirPublicKey) {
    return signumCrypto.aesSharedEncrypt(plaintext, myPrivateKey, theirPublicKey);
  }

  public static byte[] aesEncrypt(byte[] plaintext, byte[] myPrivateKey, byte[] theirPublicKey, byte[] nonce) {
    return signumCrypto.aesSharedEncrypt(plaintext, myPrivateKey, theirPublicKey, nonce);
  }

  public static byte[] aesDecrypt(byte[] ivCiphertext, byte[] myPrivateKey, byte[] theirPublicKey) {
    return signumCrypto.aesSharedDecrypt(ivCiphertext, myPrivateKey, theirPublicKey);
  }

  public static byte[] aesDecrypt(byte[] ivCiphertext, byte[] myPrivateKey, byte[] theirPublicKey, byte[] nonce) {
    return signumCrypto.aesSharedDecrypt(ivCiphertext, myPrivateKey, theirPublicKey, nonce);
  }

  public static byte[] getSharedSecret(byte[] myPrivateKey, byte[] theirPublicKey) {
    return signumCrypto.getSharedSecret(myPrivateKey, theirPublicKey);
  }

  public static String rsEncode(long id) {
    return signumCrypto.rsEncode(SignumID.fromLong(id));
  }

  public static long rsDecode(String rsString) {
    return signumCrypto.rsDecode(rsString).getSignedLongId();
  }
}

File: src/main/kotlin/brs/crypto/Crypto.kt
package brs.crypto

import burst.kit.crypto.BurstCrypto
import burst.kit.entity.BurstID
import org.bouncycastle.jcajce.provider.digest.MD5

import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

internal val burstCrypto = BurstCrypto.getInstance()

object Crypto {
    fun sha256(): MessageDigest {
        return burstCrypto.sha256
    }

    fun shabal256(): MessageDigest {
        return burstCrypto.shabal256
    }

    fun ripeMD160(): MessageDigest {
        return burstCrypto.ripeMD160
    }

    fun md5(): MessageDigest { // TODO unit test
        return try {
            MessageDigest.getInstance("MD5") // TODO burstkit4j integration
        } catch (e: NoSuchAlgorithmException) {
            MD5.Digest()
        }
    }

    fun getPublicKey(secretPhrase: String): ByteArray {
        return burstCrypto.getPublicKey(secretPhrase)
    }

    fun getPrivateKey(secretPhrase: String): ByteArray {
        return burstCrypto.getPrivateKey(secretPhrase)
    }
}

fun ByteArray.verifySignature(signature: ByteArray, publicKey: ByteArray, enforceCanonical: Boolean): Boolean {
    return burstCrypto.verify(signature, this, publicKey, enforceCanonical)
}

fun ByteArray.signUsing(secretPhrase: String): ByteArray {
    return burstCrypto.sign(this, secretPhrase)
}

fun Long.rsEncode(): String {
    // TODO don't construct BurstID
    return burstCrypto.rsEncode(BurstID.fromLong(this))
}

fun String.rsDecode(): Long {
    // TODO don't construct BurstID
    return burstCrypto.rsDecode(this).signedLongId
}

fun Long.rsVerify(): Boolean {
    // TODO is this method necessary? Is it possible that this will ever fail?
    return this == this.rsEncode().rsDecode()
}
